### ISSUE TITLE
微调Reranker问题

### ISSUE THREAD

**Issue by mfxss** [NovaSearch-Team/RAG-Retrieval#105]
相同的数据与BGE微调对比，偶尔topN召回比例会相差80%以上，出现概率大约为50%。
相同的数据训练第二轮，topN召回正例百分比会从90%骤降到10%。以下为我的训练配置。
分别对比了pairwise和listwie损失，都会出现这种情况，相同数据下BGE不会。
我的query大概为3000条，每条query对应2条正例，70条反例。
```
model_type: "bert_encoder"
num_labels: 1
query_format: "{}"
document_format: "{}"
seq: "\n"
special_token : "\n帮我找出与输入相近的文本"

## Grouped Dataset
train_dataset: ""
train_dataset_type: "grouped"
train_label_key: "label"
train_group_size: 10
shuffle_rate: 0.0
max_len: 5000
val_dataset: "../../../example_data/grouped_reranker_eval_data.jsonl"
val_dataset_type: "grouped"
val_label_key: "label"
loss_type: "listwise_ce"  # "pairwise_ranknet" or "listwise_ce"


# Training
output_dir: ""

## Model Saving
save_on_epoch_end: 1
num_max_checkpoints: 10

## Hyperparameters
epochs: 10
lr: 5e-5
batch_size: 2
seed: 42
warmup_proportion: 0.1
stable_proportion: 0.0
gradient_accumulation_steps: 2
mixed_precision: fp16
```

**Comment by NLPJCL** [comment]
@mfxss 辛苦也可以贴下bge的训练脚本，我们排查下。

**Comment by NLPJCL** [comment]
> > [@mfxss](https://github.com/mfxss) 辛苦也可以贴下bge的训练脚本，我们排查下。
> 
> CUDA_VISIBLE_DEVICES=2,3 torchrun --nproc_per_node 2 -m FlagEmbedding.finetune.reranker.encoder_only.base --model_name_or_path /mnt/nvme0n1/bge-reranker-v2-m3 --cache_dir ./cache/model --train_data /home/gkg/embed_train.jsonl --cache_path ./cache/data --train_group_size 8 --query_max_len 3000 --passage_max_len 3000 --pad_to_multiple_of 8 --knowledge_distillation True --output_dir /mnt/nvme0n1/bge-reranker-train-flag/bge-reranker-v2-m3-dfkj-all-0613 --overwrite_output_dir --learning_rate 6e-5 --fp16 --num_train_epochs 5 --per_device_train_batch_size 16 --gradient_accumulation_steps 1 --dataloader_drop_last True --warmup_ratio 0.1 --gradient_checkpointing --weight_decay 0.01 --logging_steps 1 --save_steps 85 --deepspeed ./examples/finetune/ds_stage0.json


参考这里的逻辑，实际上你用bge训练只用了8个负例。
https://github.com/FlagOpen/FlagEmbedding/blob/875fd4ffcb894313a918e8bb0c9b7ead82f37e64/FlagEmbedding/abc/finetune/reranker/AbsDataset.py#L146

**Comment by Ziyang1060** [comment]
> 对于正例反例的逻辑，我是将bge的训练数据，按正例-反例的顺序填入hits字段来进行处理的，不知道是否hits内的label需要按从大到小进行排序，是不是这里对结果有影响。

不需要按照label大小排序，但如果你的正例很少，hits字段不能直接将72个文档（2个正例+70负例）放进去。

**Comment by mfxss** [comment]
> > 对于正例反例的逻辑，我是将bge的训练数据，按正例-反例的顺序填入hits字段来进行处理的，不知道是否hits内的label需要按从大到小进行排序，是不是这里对结果有影响。
> 
> 不需要按照label大小排序，但如果你的正例很少，hits字段不能直接将72个文档（2个正例+70负例）放进去。

能从理论上解释吗，因为有时候效果会超过bge，有时候又变的很差。我需要控制负例在多少以内比较好呢

**Comment by NLPJCL** [comment]
> > > [@mfxss](https://github.com/mfxss) 辛苦也可以贴下bge的训练脚本，我们排查下。
> > 
> > 
> > CUDA_VISIBLE_DEVICES=2,3 torchrun --nproc_per_node 2 -m FlagEmbedding.finetune.reranker.encoder_only.base --model_name_or_path /mnt/nvme0n1/bge-reranker-v2-m3 --cache_dir ./cache/model --train_data /home/gkg/embed_train.jsonl --cache_path ./cache/data --train_group_size 8 --query_max_len 3000 --passage_max_len 3000 --pad_to_multiple_of 8 --knowledge_distillation True --output_dir /mnt/nvme0n1/bge-reranker-train-flag/bge-reranker-v2-m3-dfkj-all-0613 --overwrite_output_dir --learning_rate 6e-5 --fp16 --num_train_epochs 5 --per_device_train_batch_size 16 --gradient_accumulation_steps 1 --dataloader_drop_last True --warmup_ratio 0.1 --gradient_checkpointing --weight_decay 0.01 --logging_steps 1 --save_steps 85 --deepspeed ./examples/finetune/ds_stage0.json
> 
> 因为这个train_group_size参数只设置了8，参考这里的逻辑，实际上你用bge训练只用了8个负例。 https://github.com/FlagOpen/FlagEmbedding/blob/875fd4ffcb894313a918e8bb0c9b7ead82f37e64/FlagEmbedding/abc/finetune/reranker/AbsDataset.py#L146

@mfxss 我觉得主要是这里的不公平对比，对结果影响比较大，你可以设置成一样看看。

**Comment by mfxss** [comment]
> > 能从理论上解释吗，因为有时候效果会超过bge，有时候又变的很差。我需要控制负例在多少以内比较好
> 
> 这和框架处理数据集的逻辑有关：
> 
> > 你提到「每条query对应2条正例，70条反例」，而框架支持的逻辑是：从hits中按照串行顺序取出train_group_size个文档来组成一个优化group，以计算 pairwise 或 listwise loss。这可能与你当前的数据组织形式不兼容。
> 
> 假设hits长度为72，group大小为10，因为仅有2个正例，所以会有很多训练group里面没有任何正例，这将导致无法学习。
> 
> 我建议你可以先和bge微调参数对齐，保证hits的长度为8，然后进行正例和负例的自由搭配：
> 
> ```
> {"query": xxx, "hits(8个文档，其中包括1个正例和7个负例)": [{"content": str, "label": int|float}, ...]}
> {"query": xxx, "hits(8个文档，其中包括1个正例和7个负例)": [{"content": str, "label": int|float}, ...]}
> {"query": xxx, "hits(8个文档，其中包括1个正例和7个负例)": [{"content": str, "label": int|float}, ...]}
> ```

好的，我会修改数据进行对比实验

**Comment by mfxss** [comment]
> [@mfxss](https://github.com/mfxss) 有最新进展嘛

有的，将bge的group size改为80，效果变化不大，但是我们的框架还会偶发此错误，选择pair wise损失函数目前没有发现此种情况。

**Comment by NLPJCL** [comment]
> > [@mfxss](https://github.com/mfxss) 有最新进展嘛
> 
> 有的，将bge的group size改为80，效果变化不大，但是我们的框架还会偶发此错误，选择pair wise损失函数目前没有发现此种情况。

感谢反馈～我会尽快排查下。
就是相同的数据，我们的框架在pair wise损失下正常，在listwise loss下会出现召回率下降的问题哈？

**Comment by NLPJCL** [comment]
后沟通调查出来差距的原因：
bge的loss是多任务学习，listwise和蒸馏loss一起学习。
rag-retrieval，是只有蒸馏loss在学习。
如果只是蒸馏场景，不如point-wise直接学习。
